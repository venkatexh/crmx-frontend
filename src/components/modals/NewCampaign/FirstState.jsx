import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import "../../../sass/components/modals/newCampaign.scss";
import { saveTags } from "../../../redux/actions/account/saveTags";

const FirstState = ({
  name,
  tags,
  handleTagAddition,
  handleNameChange,
  handleStateChange,
}) => {
  const [errorMessage, setErrorMessage] = useState("");
  const state = useSelector(({ loggedUser, campaigns, tags }) => ({
    loggedUser,
    campaigns,
    tags,
  }));
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(saveTags(state.loggedUser.id, state.loggedUser.organization));
  }, []);

  const handleButtonClick = (e) => {
    e.preventDefault();
    if (name === "" || !name.trim()) {
      setErrorMessage("Please enter campaign name.");
    } else if (tags.length === 0) {
      setErrorMessage("Please select at least one tag.");
    } else {
      handleStateChange();
    }
  };

  return (
    <div className={"newCampaign"}>
      <div className={"header"}>Create new campaign</div>
      <form className={"form"}>
        <div>
          <label>Give a unique name to your campaign</label>
          <input
            className={"modalInput inputLarge"}
            value={name}
            placeholder={"Enter campaign name"}
            onChange={(e) => handleNameChange(e)}
          />
        </div>
        <div className={"tagsInput"}>
          <label>Add tags to your campaign</label>
          <div className={"tagsList"}>
            <div className={"userTagsList"}>
              <div className={"listWrapper"}>
                {state.tags.map((tag) => {
                  return (
                    <div
                      className={`${
                        tags.filter((t) => t.id === tag._id).length > 0
                          ? "userTagSelected"
                          : "userTag"
                      }`}
                      key={tag._id}
                      onClick={() => handleTagAddition(tag)}
                    >
                      {tag.name}
                    </div>
                  );
                })}
              </div>
            </div>
            <div className={"selectedTagsList"}>
              {tags.map((tag) => {
                return (
                  <div key={tag.id} className={"selectedTag"}>
                    {tag.name}
                  </div>
                );
              })}
            </div>
          </div>
        </div>
        <div className={"btnContainer"}>
          <div className={"errorMessage"}>{errorMessage}</div>
          <button className={"button"} onClick={handleButtonClick}>
            Next
          </button>
        </div>
      </form>
    </div>
  );
};

export default FirstState;
