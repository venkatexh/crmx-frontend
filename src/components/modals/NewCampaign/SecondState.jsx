import {useEffect, useState} from "react";
import "../../../sass/components/modals/newCampaign.scss";
import Axios from "axios";
import hostHeader from "../../../config/host";
import {useDispatch, useSelector} from "react-redux";
import {updateCampaigns} from "../../../redux/actions/account/updateCampaigns";

const SecondState = ({
                       name,
                       tags,
                       subject,
                       from,
                       text,
                       html,
                       handleTextChange,
                       handleHtmlChange,
                       handleStateChange,
                       handleSubjectChange,
                       handleFromChange,
                       handlePrevState,
                     }) => {
  const [errorMessage, setErrorMessage] = useState("");
  const [loading, setLoading] = useState(false);
  const [emailType, setEmailType] = useState(0);
  const [sendTo, setSendTo] = useState([]);

  const dispatch = useDispatch();

  const state = useSelector(({loggedUser, campaigns}) => ({
    loggedUser,
    campaigns,
  }));

  useEffect(() => {
    setLoading(true);
    for (const tag of tags) {
      Axios.get(`${hostHeader.url}/api/tag/${tag.id}/contacts`)
        .then((res) => {
          setSendTo((prevState) => [...prevState, ...res.data]);
          setLoading(false);
        })
        .catch((err) => {
          console.log(err.message);
          setErrorMessage("Something went wrong! please try again.");
          setLoading(false);
        });
    }
  }, [tags]);

  const handleButtonClick = async (e) => {
    e.preventDefault();
    if (subject === "" || !subject.trim()) {
      setErrorMessage("Please enter the subject.");
    } else if (emailType === 0 && (text === "" || !text.trim())) {
      setErrorMessage("Please enter email text.");
    } else if (emailType === 1 && (html === "" || !html.trim())) {
      setErrorMessage("Please enter the HTML code.");
    } else {
      setLoading(true);
      const campaign = {
        name,
        tags: tags.map((tag) => tag.id),
        subject,
        from,
        text,
        html,
        sentTo: sendTo,
      };
      Axios.post(
        `${hostHeader.url}/api/user/${state.loggedUser.id}/campaigns`,
        campaign
      )
        .then((res) => {
          if (res.status === 200) {
            dispatch(updateCampaigns([...state.campaigns, res.data]));
            handleStateChange();
          }
        })
        .catch((err) => {
          console.log(err.message);
          setErrorMessage("Something went wrong! please try again.");
        });
    }
  };
  if (loading) {
    return (
      <div className={"campaignSubmitLoading"}>
        <img
          src={"/loaders/comp_loader.gif"}
          alt={"loader"}
          className={"loader"}
        />
      </div>
    );
  } else {
    return (
      <div className={"newCampaign"}>
        <div className={"header"}>Create new campaign</div>
        <form className={"form"}>
          <div>
            <label>Create an email for your campaign</label>
            <input
              className={"modalInput inputLarge"}
              value={subject}
              placeholder={"Enter email subject"}
              onChange={(e) => {
                setErrorMessage("");
                handleSubjectChange(e);
              }}
            />
          </div>
          <div>
            <input
              className={"modalInput inputLarge"}
              value={from}
              placeholder={"From"}
              onChange={(e) => {
                setErrorMessage("");
                handleFromChange(e);
              }}
            />
          </div>
          <div className={"selectionGroup"}>
            <div
              className={`${emailType === 0 ? "selected" : ""} selection left`}
              onClick={() => {
                setErrorMessage("");
                setEmailType(0);
              }}
            >
              Text
            </div>
            <div
              className={`${emailType === 1 ? "selected" : ""} selection right`}
              onClick={() => {
                setErrorMessage("");
                setEmailType(1);
              }}
            >
              HTML
            </div>
          </div>
          {emailType === 0 ? (
            <div>
              <textarea
                rows={10}
                className={"emailText"}
                value={text}
                onChange={(e) => handleTextChange(e)}
              />
            </div>
          ) : (
            <div className={"tagsInput"}>
              <div className={"tagsList"}>
                <div className={"selectedTagsList"}></div>
                <div className={"userTagsList"}></div>
              </div>
            </div>
          )}
          <div className={"btnContainer"}>
            <div className={"errorMessage"}>{errorMessage}</div>
            <button className={"button"} onClick={handleButtonClick}>
              Create
            </button>
          </div>
        </form>
        <div className={"goBack"} onClick={handlePrevState}>
          go back
        </div>
      </div>
    );
  }
};

export default SecondState;
