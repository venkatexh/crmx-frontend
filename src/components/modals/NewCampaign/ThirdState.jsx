import hostHeader from "../../../config/host";
import Axios from "axios";
import {useState} from "react";

const ThirdState = ({campaignId}) => {

  const [success, setSuccess] = useState(false);
  const [error, setError] = useState(true);
  const [loading, setLoading] = useState(false);

  const sendCampaign = async () => {
    setLoading(true)
    Axios.post(`${hostHeader.url}/api/email/send`,
      {campaignId}).then(res => {
        if(res.status === 200) {
          setSuccess(true);
          setLoading(false);
        } else {
          setError(true);
          setLoading(false);
        }
    }
  )}

  const divToRender = () => {
    if(loading) {
      return <div className={"campaignSubmitLoading"}><img src={'/loaders/comp_loader.gif'} alt={'loader'} /></div>
    }
    else if(error) {
      return <div className={'campaignSubmitLoading'}>Oops. Something went wrong! Please try again.</div>
    } else if(success) {
      return <div className={'campaignSubmitLoading'}>Campaign was sent successfully!</div>
    } else {
      return <div className={"newCampaign"}>
        <div className={"success"}>
          <img
            src={"/icons/checked.png"}
            alt={"success"}
            className={"successIcon"}
          />
          <div className={"message"}>Your campaign was created!</div>
        </div>
        <div className={"buttonsGroup"}>
          <button className={"button send"} onClick={sendCampaign}>Send Now</button>
          <button className={"button schedule"}>Schedule</button>
        </div>
      </div>
    }
  }
  return divToRender();
};
export default ThirdState;
