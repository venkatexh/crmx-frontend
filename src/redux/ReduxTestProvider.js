import { Provider } from "react-redux";

const ReduxTestProvider = ({ children, reduxStore }) => (
  <Provider store={reduxStore}>{children}</Provider>
);

export default ReduxTestProvider;
