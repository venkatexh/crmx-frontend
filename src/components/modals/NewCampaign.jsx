import { useState } from "react";
import FirstState from "./NewCampaign/FirstState";
import SecondState from "./NewCampaign/SecondState";

const NewCampaign = () => {
  const [name, setName] = useState("");
  const [tags, setTags] = useState([]);
  const [subject, setSubject] = useState("");
  const [text, setText] = useState("");
  const [html, setHtml] = useState("");
  const [currentState, setCurrentState] = useState(0);

  const handleTagAddition = (tag) => {
    if (tags.filter((t) => t.id === tag._id).length > 0) {
      setTags(tags.filter((found) => found.id !== tag._id));
    } else {
      setTags((prev) => [...prev, { id: tag._id, name: tag.name }]);
    }
  };

  const stateMap = new Map();
  stateMap.set(
    0,
    <FirstState
      name={name}
      tags={tags}
      handleNameChange={(e) => setName(e.target.value)}
      handleStateChange={() => setCurrentState(currentState + 1)}
      handleTagAddition={handleTagAddition}
    />
  );

  stateMap.set(
    1,
    <SecondState
      subject={subject}
      text={text}
      html={html}
      name={name}
      tags={tags}
      handleSubjectChange={(e) => setSubject(e.target.value)}
      handleTextChange={(e) => setText(e.target.value)}
      handleHtmlChange={(e) => setHtml(e.target.value)}
      handleStateChange={() => setCurrentState(currentState + 1)}
      handlePrevState={() => setCurrentState(currentState - 1)}
    />
  );

  const componentToRender = () => {
    return stateMap.get(currentState);
  };

  return <div>{componentToRender()}</div>;
};

export default NewCampaign;
