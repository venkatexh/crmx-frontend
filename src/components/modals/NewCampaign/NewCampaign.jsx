import { useState } from "react";
import FirstState from "./FirstState";
import SecondState from "./SecondState";
import ThirdState from "./ThirdState";

const NewCampaign = () => {
  const [name, setName] = useState("");
  const [tags, setTags] = useState([]);
  const [subject, setSubject] = useState("");
  const [text, setText] = useState("");
  const [html, setHtml] = useState("");
  const [currentState, setCurrentState] = useState(0);

  const handleTagAddition = (tag) => {
    if (tags.filter((t) => t.id === tag._id).length > 0) {
      setTags(tags.filter((found) => found.id !== tag._id));
    } else {
      setTags((prev) => [...prev, { id: tag._id, name: tag.name }]);
    }
  };

  const stateMap = new Map();
  stateMap.set(
    0,
    <FirstState
      name={name}
      tags={tags}
      handleNameChange={(e) => setName(e.target.value)}
      handleStateChange={() => setCurrentState(currentState + 1)}
      handleTagAddition={handleTagAddition}
    />
  );

  stateMap.set(
    1,
    <SecondState
      subject={subject}
      text={text}
      html={html}
      name={name}
      tags={tags}
      handleSubjectChange={(e) => setSubject(e.target.value)}
      handleTextChange={(e) => setText(e.target.value)}
      handleHtmlChange={(e) => setHtml(e.target.value)}
      handleStateChange={() => setCurrentState(currentState + 1)}
      handlePrevState={() => setCurrentState(currentState - 1)}
    />
  );

  stateMap.set(2, <ThirdState />);

  const componentToRender = () => {
    return stateMap.get(currentState);
  };

  return <>{componentToRender()}</>;
};

export default NewCampaign;
