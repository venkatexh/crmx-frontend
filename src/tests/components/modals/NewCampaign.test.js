import { shallow } from "enzyme";
import React from "react";
import NewCampaign from "../../../components/modals/NewCampaign/NewCampaign";
import FirstState from "../../../components/modals/NewCampaign/FirstState";
import SecondState from "../../../components/modals/NewCampaign/SecondState";
import ThirdState from "../../../components/modals/NewCampaign/ThirdState";
import ReduxTestProvider from "../../../redux/ReduxTestProvider";

it("should be new campaign modal component snapshot", () => {
  expect(shallow(<NewCampaign />)).toMatchSnapshot();
});

it("should be first state component snapshot", () => {
  expect(
    shallow(
      <ReduxTestProvider>
        <FirstState />
      </ReduxTestProvider>
    )
  ).toMatchSnapshot();
});

it("should be second state component snapshot", () => {
  expect(
    shallow(
      <ReduxTestProvider>
        <SecondState />
      </ReduxTestProvider>
    )
  ).toMatchSnapshot();
});

it("should be third state component snapshot", () => {
  expect(shallow(<ThirdState />)).toMatchSnapshot();
});
