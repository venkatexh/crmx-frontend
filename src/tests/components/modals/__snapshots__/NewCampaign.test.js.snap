// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be first state component snapshot 1`] = `ShallowWrapper {}`;

exports[`should be new campaign modal component snapshot 1`] = `ShallowWrapper {}`;

exports[`should be second state component snapshot 1`] = `ShallowWrapper {}`;

exports[`should be third state component snapshot 1`] = `ShallowWrapper {}`;
